src/libutil/c/nix_api_util.cc

```diff
@@ -0,0 +1,151 @@
+    }
+    // unreachable
+}
+
+nix_err nix_set_err_msg(nix_c_context * context, nix_err err, const char * msg)
+{
+    if (context == nullptr) {
+        // todo last_err_code
+        throw nix::Error("Nix C api error: %s", msg);
+    }
+    context->last_err_code = err;
+    context->last_err = msg;
+    return err;
+}
+
+const char * nix_version_get()
+{
+    return PACKAGE_VERSION;
+}
+
+// Implementations
+nix_err nix_setting_get(nix_c_context * context, const char * key, char * value, int n)
+{
+    if (context)
+        context->last_err_code = NIX_OK;
+    try {
+        std::map<std::string, nix::AbstractConfig::SettingInfo> settings;
+        nix::globalConfig.getSettings(settings);
+        if (settings.contains(key))
+            return nix_export_std_string(context, settings[key].value, value, n);
+        else {
+            return nix_set_err_msg(context, NIX_ERR_KEY, "Setting not found");
+        }
+    }
+    NIXC_CATCH_ERRS
+}
+
+nix_err nix_setting_set(nix_c_context * context, const char * key, const char * value)
+{
+    if (context)
+        context->last_err_code = NIX_OK;
+    if (nix::globalConfig.set(key, value))
+        return NIX_OK;
+    else {
+        return nix_set_err_msg(context, NIX_ERR_KEY, "Setting not found");
+    }
+}
+
+nix_err nix_libutil_init(nix_c_context * context)
+{
+    if (context)
+        context->last_err_code = NIX_OK;
+    try {
+        nix::initLibUtil();
+        return NIX_OK;
+    }
+    NIXC_CATCH_ERRS
+}
+
+const char * nix_err_msg(nix_c_context * context, const nix_c_context * read_context, unsigned int * n)
+{
+    if (context)
+        context->last_err_code = NIX_OK;
+    if (read_context->last_err) {
+        if (n)
+            *n = read_context->last_err->size();
+        return read_context->last_err->c_str();
+    }
+    nix_set_err_msg(context, NIX_ERR_UNKNOWN, "No error message");
+    return nullptr;
+}
+
+nix_err nix_err_name(nix_c_context * context, const nix_c_context * read_context, char * value, int n)
+{
+    if (context)
+        context->last_err_code = NIX_OK;
+    if (read_context->last_err_code != NIX_ERR_NIX_ERROR) {
+        return nix_set_err_msg(context, NIX_ERR_UNKNOWN, "Last error was not a nix error");
+    }
+    return nix_export_std_string(context, read_context->name, value, n);
+}
+
+nix_err nix_err_info_msg(nix_c_context * context, const nix_c_context * read_context, char * value, int n)
+{
+    if (context)
+        context->last_err_code = NIX_OK;
+    if (read_context->last_err_code != NIX_ERR_NIX_ERROR) {
+        return nix_set_err_msg(context, NIX_ERR_UNKNOWN, "Last error was not a nix error");
+    }
+    return nix_export_std_string(context, read_context->info->msg.str(), value, n);
+}
+
+nix_err nix_err_code(const nix_c_context * read_context)
+{
+    return read_context->last_err_code;
+}
+
+// internal
+nix_err nix_export_std_string(nix_c_context * context, const std::string_view str, char * dest, unsigned int n)
```

Comment:
Done in https://github.com/NixOS/nix/pull/8699/commits/4c01714d02231a3bc5d105c49299371e8b992102

I'm thinking that maybe something like `nix_get_string_callback` (instead `nix_observe_string`) could make more sense, but I don't have a strong preference.

Reply:
Oh I forgot to mention that the same signature can also be used for returning sets and lists of strings.

Reply:
@roberth Could you clarify what you mean? Not sure if you mean that the name `nix_observe_string` is ok, or if you are requesting some other change

Reply:
This was in response to the naming.
I don't think we have a specific enough use case for multiple calls. The closest thing is "attrNames", but there the integer based indexing works rather well, because the size is known in advance, and we can simultaneously retrieve the values.

Reply:
Sorry for the uncertainty. `nix_get_string_callback` sounds pretty good actually, considering that we might not have a need to call it multiple times anywhere.

Reply:
All good, no need to apologize :)
I renamed it to `nix_get_string_callback`